onnxruntime/core/providers/webgpu/tensor/transpose.cc

```diff
@@ -86,16 +127,43 @@ Status Transpose::ComputeInternal(ComputeContext& context) const {
   TensorShape output_shape(output_dims);
   auto* output_tensor = context.Output(0, output_shape);
 
+  InlinedVector<int64_t> new_shape{};
+  InlinedVector<int64_t> new_perm{};
+  SqueezeShape(input_shape.GetDims(), *p_perm, new_shape, new_perm);
+  const auto channels_last = new_perm == InlinedVector<int64_t>({2, 3, 1});
+  const auto channels_first = new_perm == InlinedVector<int64_t>({3, 1, 2});
+  const auto use_shared = (new_shape.size() == 2 && new_perm[0] > new_perm[1]) || channels_last || channels_first;
+  auto new_input_shape = use_shared ? new_shape : input_shape;
```

Comment:
try to avoid unnecessary assignments (when use_shared == true, `new_input_shape` will be assigned again below)